### ISSUE TITLE
ViewableData::setFailover() creates artefacts that break hasMethod()

### ISSUE THREAD

**Issue by unclecheese** [silverstripe/silverstripe-framework#9140]
## Scenario:

* my-block-page (BlockPage)
    * my-regular-page (Page

`GET /my-block-page/my-regular-page`

* `ModelAsController::handleRequest()` gets instance of `BlockPage`
* `BlockPage` does not have a `BlockPageController`. Nested controller falls back on `PageController`
* `PageController::create($blockPage)` sets `failover` to `BlockPage`.
* Extra methods from `BlockPage` are loaded into the **static** property `$extra_methods['PageController']`
* `my-regular-page` comes back with *another* instance of `PageController` using `$myPage` as its record.
* `PageController::setFailover($myPage)` should now clear out all the artefacts from `BlockPage`, (`removeMethodsFrom('failover')`). But it doesn't, because `$this->failover` doesn't exist:

https://github.com/silverstripe/silverstripe-framework/blob/4/src/View/ViewableData.php#L170

In short, this appears to be inconsistency between the static and instance level state.

Possible fix would be, instead of checking `$this->failover` exists, check if there are any methods in `static::class` where `$methodInfo['property'] === 'failover'`.

That would be a non-trivial performance hit, though.

**Comment by tractorcow** [comment]
I can confirm this is the case too; I found this issue with Elemental and the ElementController breaking on different content blocks reporting the wrong hasMethod().

Simple explanation:

Multiple controllers of the same class will share one cache for methods. If you generate a list of controllers with different items, only the LAST created controller will be correct for the failover it is assigned. All other controllers (still with their own failover) will think they have the methods of the last.

Suggested solution: Cache failover methods on both the controller parent, and failover class names, instead of only the controller name.

I suggest adding a new instance property, which is method key. setFailover would assign this key, and any CustomMethods code which touches the $extra_methods static would respect it.
